File: Java/75.sort colors.two pointers.java
class Solution {
    public void sortColors(int[] nums) {
        int len = nums.length;
        if (len < 2) {
            return;
        }
        int zero = 0;
        int two = len;
        int i = 0;
        while (i < two) {
            if (nums[i] == 0) {
                swap(nums, i, zero);
                i++;
                zero++;
            } else if (nums[i] == 1) {
                i++;
            } else {
                two--;
                swap(nums, i, two);
            }
        }
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
// time complexity : faster 100% O(n)
// space complexity : faster 74.58% O(1)

File: cpp/75.Sort-Colors.TwoPointers.cpp
// 100% faster
// O(n), O(1)

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int p{0}, q{(int)nums.size() - 1};
        for(int i{}; i<=q; i++){
            if(nums[i] == 0){
                swap(nums[i], nums[p]);
                p++;
            }
            else if(nums[i] == 2){
                swap(nums[i], nums[q]);
                q--;
                i--;
            }
        }
    }
};
